components/datetime/src/format/neo.rs

```diff
@@ -350,9 +350,9 @@ size_test!(
 ///     [
 ///         (7, 10, Part::ERROR), // mon
 ///         (11, 14, Part::ERROR), // M11
-///         (23, 25, Part::ERROR), // ce
-///         (42, 44, Part::ERROR), // AM
-///         (45, 50, Part::ERROR), // +0000
+///         (23, 36, Part::ERROR), // (era unknown)
```

Comment:
@sffc I don't really think this `assert_try_writeable_parts_eq` test is pulling its weight as a docs test; these values are brittle and also largely uninteresting to the docs reader

Reply:
It's not brittle, it changes when the output or the errors change, which are client-observable, relevant changes. I think this is a very valuable test, it's one of the few ones that demonstrate how ranges of the output are marked as errors.

Reply:
This should btw not be `(era unknown)`. This is missing *names*, we should still know which era it is, and there should be a better fallback value. "ce" was a good fallback value for the `ce` era. "era unknown" sounds like a `MissingInputField`, which results in `{G}`, like other missing inputs.

Reply:
@robertbastian I think that could be achieved by testing that a single range is marked with an error. It's using a testing-only macro, I don't actually think that's super helpful when it comes to understanding how ranges work. 

Personally I think we should have demonstrations of _working_ ranges in this crate and errorful ones in the Writeable crate.

Reply:
I'd love to test with with an actual write-to-parts sink, but we don't currently have one.

Reply:
re: era unknown: Do you have thoughts on how to solve that? Because currently the era is no longer present here.

 - Produce an unconditional formatting era
 - Include an index-to-code map for the calendar/DateTimeInput (this may monomorphize-bloat a lot of this code)

Reply:
I would also be fine with producing `{G}` here, I think that's perfectly OK.

Reply:
But the era is present. `GetField` gives you something, if it didn't then it's a missing field error (only that should produce `{G}`). If `GetField` gives you `::Index(1)`, the calendar knows what that means (it *really* should, if it doesn't then I disagree with this whole PR).

Reply:
Okay, how do you think this should be achieved in code? Like, I don't like it either, but couldn't come up with a clean way to solve it. The easiest might be to have a fallback `&'static str` era next to the index, or a TinyStr. I can do that if you'd like.

Reply:
Add a method to the calendar trait that resolves a `YearEra` and an `Option<&EraNamesStructThingV1>` to a string. That way you can also store japanese era keys in the singleton, and can store the names in a list.

Reply:
The calendar is not persisted to this point, and persisting it would trigger monomorphization. This code has been deliberately crafted in a way to not monomorphize too much; it's a lot of code.

That's also the architectural problem with persisting Japanese era keys to this point.

Reply:
You could create a static names data struct with "ce" and "ad" and format with that if names are missing.

Reply:
The calendar could return a `VarZeroVec<str>` which are the era keys. Gregorian would return a static VZV containing "ce" and "ad", japanese would load this from data. The names then contain a VZV of the same length, which are the values. If the values are missing, you use the keys. This basically splits the ZeroMap, deduplicating the keys, and it keeps the era identifiers human-readable (`ce` is much better than `::Index(1)`).

Reply:
It could still fail if the index is unsuspected, or if the formats don't match, or if the field length isn't available (which would be a bunch of work to detect in advance to figure out if a fallback is needed, since we don't know the calendar at this stage). I don't think that's a good fallback.

I'd rather just add a fallback tinystr to the FormattableEra enum.

Reply:
I'd like to know what @sffc thinks of how to achieve this.

Reply:
> I'd rather just add a fallback tinystr to the FormattableEra enum.

That's what I'm proposing. And the calendar gives you a VZV to go from tinystr to index, for lookup in names. If that fails, report an error and format the tinystr.

Reply:
@robertbastian "And the calendar gives you a VZV to go from tinystr to index" is the hard part. I'd rather just do the fallback tinystr with no additional VZV.

Reply:
Why is this hard? This can happen way before `GetField`, at some point the calendar is still there.

Reply:
I definitely like that our current identifiers are fast, but adding "era name map" to the date time input model is not a trivial proposal.

Reply:
> Why is this hard? This can happen way before `GetField`, at some point the calendar is still there.

The calendar is not available to the datetime formatter at all (except as a bare phantom type for the typed version). The datetime is, and we have the YearInput/etc stuff to prefetch all necessary data before formatting.

Reply:
The end result of this would basically be adding this vector to YearInfo, or potentially to some type wrapping this.

Which is heavier than my not-ideal-but-workable proposal of just passing along a fallback string. A fallback string is definitely less heavy than passing the entire map around each time you format.

Reply:
There may be ways to make the thing you are proposing work, but I can't see a clear way to do so, and I'm not signing up for it, let's land this and if you wish to further experiment, go for it.

Reply:
(I already fixed the fallbacking using the scheme I proposed above. I had considered it earlier and mostly didn't want to go too far down that route without some validation that we cared about fallback, which I got)

Reply:
I don't want to land this like this. I think communicating eras as indices over the `GetField` API is bad, and it creates an issue with missing names. Adding a fallback string doesn't solve the problem, because there's nothing enforcing that it matches the name behind the index.

Reply:
> Adding a fallback string doesn't solve the problem, because there's nothing enforcing that it matches the name behind the index.


I mean, whether the association between index-and-era-code is in the `YearInfo` struct constructor in the calendar implementation, or in some `get_year_map()` function in the calendar implementation it seems like it ends up living in one place (plus datagen, which wouldn't link to it in either way). So I'm not sure either is more or less brittle.